### PR TITLE
Issue 3274: (SegmentStore) Table Segment Compaction

### LINKED ISSUE #3274: Table Segment Compaction
**Problem description**
Table Segments are prone to fragmentation and require compaction.

**Problem location**
Table Segments.

**Suggestions for an improvement**

Compaction Triggers:
- Keep track of Table Segment Effective Size (sum of all Table Entry Lengths). This is done in #3273 .
- Define fragmentation as 1-TableSize/TableSegmentLength (this is a number between 0 and 1; the smaller the more compact the segment is).
- Create a trigger (in Table Config) which defines the threshold above which a compaction is triggered.

Compaction Process:
- We need to go through every Table Entry from the beginning of the Segment (beginning = the first non-truncated offset).
- We inspect each Table Entry that we encounter. 
    - If it is no longer indexed (no TableBucket or backpointer pointing to it), it can be truncated out.
    - If it is still indexed, it needs to be **safely** moved to the end of the Segment. The word **safe** is important here, as a more recent update of that TableEntry may have been recorded in the meantime - we do not want to lose new data.
- When we process enough entries, we perform a Segment Truncation (threshold configurable)
- When the fragmentation of the TableSegment reaches below the desired threshold (see above), compaction can end.

### PR DESCRIPTION
**Pull request by andreipaduroiu**
**Change log description**  
- Implemented Table Segment Compaction
- Defined two more Table Segment Attributes that aid in compaction: one for tracking utilization and one for tracking compaction progress.
- Compaction is not enabled for any segment yet (all have MIN_UTILIZATION set to 0, which means no compaction). This will change in a future PR.

**Purpose of the change**  
Fixes #3274.

**What the code does**  
- Defined new `TableAttributes`
    - `COMPACTION_OFFSET`: keeps track of the last offset that was processed for compaction. Subsequent compaction runs will resume from here. This offset is always at a `TableEntry` boundary.
    - `MIN_UTILIZATION`: this is a user-set threshold that defines when compaction should be triggered. Utilization is defined as the ratio of `ENTRY_COUNT` to `TOTAL_ENTRY_COUNT`, expressed as a percentage (0..100). The former counts how many Table Entries are active in the segment, while the latter counts all entries, including overwritten and deleted ones.
- Table Entries can now have explicit versions
    - Previously, a Table Entry's version is defined by the offset at which it resides in the Segment. However, since compaction copies entries from the beginning of the Segment to the end, this would have a side effect of changing the entry's version, which would affect conditional updates.
    - Table Entries now have an additional serialized field which stores its explicit version. 
    - New entries (via the API) will not have this field set, in which case the Entry's version will still be inferred from the offset.
    - Copied entries (from compaction) will have this field set to whatever version the entry originally had (if it ends up being copied multiple times then the original version will be kept). As such, when being read, this version will be the one used for conditional updates.
- Compaction is performed in the background by the `WriterTableProcessor` and works along these lines:
    - A number of entries are read beginning at `COMPACTION_OFFSET`. There is a maximum of 4MB to be read at any given time to prevent compaction from using too much IO at any given time.
    - Of those entries that are read, we only consider those entries that are still active in the index (so we exclude deleted or obsolete entries). These will be copied over at the end of the Table Segment, while preserving their original version.
    - A subsequent iteration of the `WriterTableProcessor` will index these copied entries and update the index as needed. There is no special treatment of these in the indexer, except for the rules defined below.
- Reindexing rules
    - When a Table Entry is copied over, we need to account for the fact that we may get a concurrent update to that Entry (via the API). As such, we want to make sure that the copied entry will not override the new value we have just added via the API (previously, anything with a higher offset would supersede anything with lower offset).
    - The `IndexWriter` has not been changed for this purpose. It will continue to update the Table Bucket Index given the inputs we provide. As such, we tweak the inputs we provide to this class, done in `WriterTableProcessor` with the help of the following classes:
        - `KeyUpdateCollection`: this (existing) class helps the `WriterTableProcessor` to organize Table Entries as they are read from the segment for indexing (by de-duping entries with the same key). This has been changed to keep Table Entries with higher versions (and not with higher offsets). It does this by calculating the correct version for each entry (based on offset and explicit version).
        - `BucketUpdate`: this (existing class) contains the state of the Table Bucket that is sent to the `IndexWriter`. It deals with a single bucket and contains both existing and updated keys. It has been changed so that a key is no longer considered updated if it has an existing key that supersedes any updates to it (same superseding rules apply as in `KeyUpdateCollection`).

- Cache invalidation

    - The compaction happens on the segment itself, however the `ContainerTableExtensionImpl` keeps a `ContainerKeyIndex`/`Cache` for recently accessed keys. This cache will need to be partially invalidated if we encounter stale values due to compaction (it auto-updates for API calls, but compaction happens from underneath it).
    - Upon reading entries, if a `StreamSegmentTruncatedException` is encountered, the cache for that bucket is invalidated and it is re-fetched from the index (and recached).
    - There is another condition for invalidating cache, but only for buckets that contain collisions (a pretty rare event).

**How to verify it**  
All existing and newly added unit tests must pass.
Since this is not yet enabled on any segments, there is no need to run system tests against this PR.

### CODE REVIEW

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/WriterTableProcessor.java [hunk 1]
```diff
@@ -238,16 +291,16 @@ public String toString() {
...
-                                                    keyUpdates.getTotalUpdateCount(), timer.getRemaining());
-                                        },
-                                        this.executor)
-                                .thenApply(ignored -> keyUpdates.getLastIndexedOffset()),
-                        this.executor);
+                                    val bucketUpdates = builders.stream().map(BucketUpdate.Builder::build).collect(Collectors.toList());
+                                    logBucketUpdates(bucketUpdates);
+                                    return this.indexWriter.updateBuckets(segment, bucketUpdates,
+                                            this.aggregator.getLastIndexedOffset(), keyUpdates.getLastIndexedOffset(),
+                                            keyUpdates.getTotalUpdateCount(), timer.getRemaining());
```
Comment (tkaitchuck) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/WriterTableProcessor.java]: keyUpdateCollection is not thread safe, but this looks like it's occurring in another thread. Perhaps keyUpdates should be passes explicitly as an argument here?

Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/WriterTableProcessor.java]: I think the best solution here is to make KeyUpdateCollection thread safe. I fixed it.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/IndexReader.java [hunk 2]
```diff
@@ -104,6 +104,26 @@ long getBucketCount(SegmentProperties segmentInfo) {
...
+        return segmentInfo.getAttributes().getOrDefault(TableAttributes.COMPACTION_OFFSET, 0L);
+    }
+
+    /**
+     * Gets the Segment Utilization (as a percentage) under which a compaction may be required.
+     *
+     * @param segmentInfo A {@link SegmentProperties} from which to extract the information.
+     * @return The result.
+     */
+    long getCompactionUtilizationThreshold(SegmentProperties segmentInfo) {
```
Comment (tkaitchuck) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/IndexReader.java]: The fact that this is returned as a long makes the calculations more complicated. 
We can have the value in the config as a percentage and here divide by 100.0 and return the double.

File: common/src/main/java/io/pravega/common/io/StreamHelpers.java [hunk 3]
```diff
@@ -54,13 +55,17 @@ public static int readAll(InputStream stream, byte[] target, int startOffset, in
...
-     * @throws IOException If unable to read from the given InputStream.
+     * @throws IOException If unable to read from the given InputStream. Throws {@link EOFException} if the number of bytes
+     * remaining in the InputStream is less than length.
      */
     public static byte[] readAll(InputStream source, int length) throws IOException {
         byte[] ret = new byte[length];
         int readBytes = readAll(source, ret, 0, ret.length);
-        Preconditions.checkArgument(readBytes == ret.length,
-                "Invalid value for length (%s). Was only able to read %s bytes from the given InputStream.", ret.length, readBytes);
+        if (readBytes < ret.length) {
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply common/src/main/java/io/pravega/common/io/StreamHelpers.java]: This is used in places where we know exactly how much we expect to read. In this particular case, we read Table Entries from the segment, for which we know the exact length. 

This is different from the method above it, which reads at most a number of bytes, and its behavior is documented in its Javadoc.

File: segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/tables/TableAttributes.java [hunk 4]
```diff
@@ -40,6 +40,17 @@
      */
     public static final UUID TOTAL_ENTRY_COUNT = new UUID(CORE_ATTRIBUTE_ID_PREFIX, TABLE_ATTRIBUTES_START_OFFSET + 3);
 
+    /**
+     * Defines an attribute that is used to store the offset of a (Table) Segment where compaction has last run at.
+     */
+    public static final UUID COMPACTION_OFFSET = new UUID(CORE_ATTRIBUTE_ID_PREFIX, TABLE_ATTRIBUTES_START_OFFSET + 4);
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/tables/TableAttributes.java]: Just because you compact (copy entries), you can't truncate right away. The Index still points to the original entry locations and if we truncate we will not be able to serve that data. A subsequent iteration of the processor will index the copied entries and it will repoint the index to the new locations. Only then it is safe to truncate this data.

File: segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/tables/TableAttributes.java [hunk 5]
```diff
@@ -40,6 +40,17 @@
...
+
+    /**
+     * Defines an attribute that is used to set the minimum utilization (as a percentage of {@link #ENTRY_COUNT} out of
+     * {@link #TOTAL_ENTRY_COUNT}) of a Table Segment below which a Table Compaction is triggered.
+     */
+    public static final UUID MIN_UTILIZATION = new UUID(CORE_ATTRIBUTE_ID_PREFIX, TABLE_ATTRIBUTES_START_OFFSET + 5);
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/tables/TableAttributes.java]: These should never be reused. If we reuse them for a different purpose, then old segments will still have the value for the previous attribute and it may be interpreted incorrectly.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/AsyncTableEntryReader.java [hunk 6]
```diff
@@ -99,6 +106,25 @@ private AsyncTableEntryReader(@NonNull EntrySerializer serializer, @NonNull Time
...
     }
 
+    /**
+     * Reads a single {@link TableEntry} from the given InputStream. The {@link TableEntry} itself is not constructed,
+     * rather all of its components are returned individually.
+     *
+     * @param input         An InputStream to read from.
+     * @param segmentOffset The Segment Offset that the first byte of the InputStream maps to. This wll be used as a Version,
+     *                      unless the deserialized segment's Header contains an explicit version.
+     * @param serializer    The {@link EntrySerializer} to ues for deserializing entries.
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/AsyncTableEntryReader.java]: fixed

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/AsyncTableEntryReader.java [hunk 7]
```diff
@@ -99,6 +106,25 @@ private AsyncTableEntryReader(@NonNull EntrySerializer serializer, @NonNull Time
...
+     * @return A {@link DeserializedEntry} that contains all the components of the {@link TableEntry}.
+     * @throws IOException If an Exception occurred while reading from the given InputStream.
+     */
+    static DeserializedEntry readEntryComponents(InputStream input, long segmentOffset, EntrySerializer serializer) throws IOException {
+        val h = serializer.readHeader(input);
+        long version = getKeyVersion(h, segmentOffset);
+        byte[] key = StreamHelpers.readAll(input, h.getKeyLength());
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/AsyncTableEntryReader.java]: I would love to do that, but my hands are a bit tied. 

This method
`public static byte[] readAll(InputStream source, int length) throws IOException`
takes in a length and returns a `byte[]` of that length. It needs to preallocate that array before it does any reading. It cannot also return the actual number of bytes that were also read, at least not without allocating yet another array and copying data to it.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java [hunk 8]
```diff
@@ -209,6 +209,26 @@ public void close() {
...
+     * an updated offset for this Bucket (i.e., as a result of a compaction), the cache will be automatically updated in
+     * the process. If the index contains an obsolete offset for this Bucket (compared to the cache), the cache value
+     * will be returned.
+     *
+     * @param segment Segment to look up the Bucket Offset for.
+     * @param keyHash The Key Hash to look up.
+     * @param timer   Timer for the operation.
+     * @return A CompletableFuture that, when completed, will contain the sought offset. If the  bucket does not exist,
+     * it will contain {@link TableKey#NOT_EXISTS}.
+     * associated with it.
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java]: Yes. I removed it.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java [hunk 9]
```diff
@@ -209,6 +209,26 @@ public void close() {
         }
     }
 
+    /**
+     * Finds the Bucket Offset for the given Key Hash directly from the index (excluding the cache). If the index contains
+     * an updated offset for this Bucket (i.e., as a result of a compaction), the cache will be automatically updated in
+     * the process. If the index contains an obsolete offset for this Bucket (compared to the cache), the cache value
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java]: Consider this scenario, for Key K:
1. We fetch K, and it was found at offset O1. Hence the cache contains K->O1.
2. We compact and move K to O2, then we truncate at offset T (O1<T<=O2). At this point, the index contains K->O2, but the cache still has K->O1.
3. We try to fetch K, but when reading the key data at offset O1, we get a `SegmentTruncatedException`. We need to get the updated position from the index (O2) and update the cache (if needed) and return the result.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerTableExtensionImpl.java [hunk 10]
```diff
@@ -220,13 +227,34 @@ public void close() {
...
+                // directly from the index if unable to find anything and there is a chance the sought key actually exists.
+                // Encountering a truncated Segment offset indicates that the Segment may have recently been compacted and
+                // we are using a stale cache value.
+                ArrayView key = builder.getKeys().get(i);
+                builder.includeResult(Futures
+                        .exceptionallyExpecting(bucketReader.find(key, offset, timer), ex -> ex instanceof StreamSegmentTruncatedException, null)
+                        .thenComposeAsync(entry -> {
+                            if (entry != null) {
+                                // We found an entry; need to figure out if it was a deletion or not.
+                                return CompletableFuture.completedFuture(maybeDeleted(entry));
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerTableExtensionImpl.java]: `maybeDeleted` returns a `TableEntry`. Either the entry that was passed, or null if the entry indicates a deletion.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableCompactor.java [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+     * Calculates the offset in the Segment where it is safe to truncate based on the current state of the Segment and
+     * the highest copied offset encountered during an index update.
+     *
+     * This method is invoked from the {@link WriterTableProcessor} after indexing. Since compaction is loosely coupled
+     * with indexing, the {@link WriterTableProcessor} does not have too many insights into what has been compacted or not;
+     * it can only decide based on the current state of the Table Segment and what it has just indexed. As such:
+     * - If recently indexed Table Entries indicate they were copied as part of a compaction, then it is safe to truncate
+     * at the highest copied offset encountered (since they are copied in order, by offset). Everything prior to this
+     * offset is guaranteed not to exist in the index anymore.
+     * - If no recently indexed Table Entry indicates it was copied as a result of compaction, then it may not be safe to
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableCompactor.java]: Correct. The truncation only is invoked if no other exceptions happen during the indexing iteration, so if we encounter such copied offsets, they must have been indexed already and the index no longer points to the original offsets. Since all these entries are copied in order of offsets, it is safe to truncate at the highest one we find).

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (fpj) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java]: Your explanation makes sense, but I'm stuck at this:

> If the index contains an obsolete offset for this Bucket (compared to the cache), the cache value will be returned.

You scenario argues about the cache containing a stale value, but this sentence talks about the cache having fresher data compared to the index.

[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java]: Yes, there are two cases:
1. Cache contains outdated information. This happens when a compaction moved an entry from the back to the front, and nothing else happened to the key (no new updates in the meantime). In this case, the correct value is loaded from the index and stored in the cache, prior to returning it.
2. Cache contains the correct value. This can happen when there has been no compaction (so index value matches cache value) or there has been a recent update to this key that is not yet reflected in the index. In this case, the value from the index is discarded and the cached value returned.

Please note that it is impossible for us to determine with exact accuracy if a key in the cache has become obsolete. If you look above (237-255) you'll see that we resort to this if one of the following happens:
- We encounter a truncated exception (we know for sure this is outdated) 
- We get a `null` entry from `bucketReader.find`. This is returned either because the entry has been deleted or if we have a multi-entry bucket (due to collisions) which had to be rearranged due to compaction (example below).

Let me explain the bucket rearrangement.
- Suppose we have 3 keys in a bucket: K1-K3.
- The bucket points to the offset of K3, and we use backpointers to link the rest: K3->K2, K2->K1.
- We compact and move K1 to the end, so the indexer will rearrange the bucket as K1->K3; K3->K2.
- The cache still points to K3, so if we look for K1 now (after reindexing), we will start at K3 and only find K3 and K2, but not K1. In this case we need to go directly to the index to get the new location of the bucket (K1) so that we may return the right result.

An alternative to this cache invalidation on demand would be to have a notification system from the indexer into the cache. But that would be unnecessarily complex, chatty and would introduce tight coupling between the components, making them much harder to test. I think that a small performance penalty (due to an unneeded retry) in a corner case is acceptable to this alternative.
